<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2014-1</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Daniel Rojo Mata
### Rodrigo Vega Vilchis
---

<h1 align="center">Programación para la física computacional</h1> 

## Práctica 1.  Programación para la física computacional


### Programación básica para la física

---

### EJERCICIOS:

**1. Caida de una pelota desde una torre:** Se deja caer una pelota desde una torre de altura $h$. Tiene velocidad inicial cero y acelera hacia abajo con la gravedad.

Escribe un programa que le pida al usuario que ingrese la altura en metros de la torre y luego calcule e imprima el tiempo en segundos hasta que la pelota toque el suelo (ignora la resistencia del aire). 
    
Después, usa tu programa para calcular el tiempo de una pelota lanzada desde una torre de $100m$ de altura.

Para este modelo el origen del sistema coordenado estara en el edificio, la aceleracion gravticacional la tomaremos como $g=9.81\frac{m}{s²}$. Por las ecuaciones de la cinematica podemos usar la ecuacion $h=v_{inicial}t+\frac{1}{2}gt²$, pero si dejamos caer la pelora desde el reposo entonces podemos decir que $v_{inicial}=0\frac{m}{s}$ y finalmene solo tenemos la ecuacion $h=\frac{1}{2}gt²$, de l cual si despejamos el tiempo tenemos que eel tiempo de caida de la pelota desde una altura $h$ es $t=\sqrt{\frac{2h}{g}}$

In [1]:
##### Pedimos al usuario la altura del edifcio en metros 
altura=float(input("Intoduce la altura del edifcio en metros:"))
#### Calculamos el tiempo de caida e imprimios el valor 
from math import *
g=9.81
tiempo=round(sqrt(2*altura/g),2)
print("El tiempo de caida es de:",tiempo,"segundos")

Intoduce la altura del edifcio en metros: 100


El tiempo de caida es de: 4.52 segundos


# Resultado: 1/1

**2. Altitud de un satélite**: se va a lanzar un satélite en una órbita circular alrededor de la Tierra de modo que orbite el planeta una vez cada $T$ segundos. 

(a) Demuestre que la altitud $h$ sobre la superficie de la Tierra que debe tener el satélite es:

<h3 align="center">$h = \biggl( {GMT^2\over4\pi^2} \biggr)^{1/3} - R$,</h3> 

donde $G=6.67\times10^{-11}\,\textrm{m}^3\,\textrm{kg}^{-1}\,\textrm{s}^{-2}$ es la constante gravitatoria de Newton, $M=5.97\times10^{24}\,$kg es la masa de la Tierra y $R=6371\,$km  es su radio.



Para demostrar esta ecuación ocuparemos la ley de gravitacion universal y el movimiento rotacional, si el satelite esta rotando al rededor del sol entonces presecencia fuerzas ficticias, y si esta en equilibrio en su orbita, esta fuerza ficticia llamada fuerza centrifuga debe ser igual en magnitud que la fuerza gravitacional que ejerce la tierra sobre el pero en sentido contrario. Entonces tenemos la ecuación \begin{align*}
    \frac{GMm_s}{r²}&=m_s\frac{v²_{orbital}}{r}\ \therefore\\
    \frac{GM}{r}&=\omega²r²=\frac{4\pi²r²}{T²}\ \Rightarrow\\
    r&=\left(\frac{GMT²}{4\pi²}\right)^{\frac{1}{3}}
\end{align*}
Donde $r$ es la distancia del satelite aal centro de la tierra, por lo que si queremos saber la altura del satelite medida desde la superficie terrestre, a la distancia $r$ hay que restarle el radio terrestre y la altura sera entonces:
\begin{equation}
    h=r-R={\left(\frac{GMT²}{4\pi²}\right)^{\frac{1}{3}}-R}_{\blacksquare}
\end{equation}


## Resultado: 0.3333/0.3333

(b) Escribe un programa que le pida al usuario que ingrese el valor deseado de T y luego calcula e imprima la altitud correcta en metros.

In [2]:
##### PRIMERO ESCRIBIMOS LA CONSTANTES EN SI
R=6371e3
G=6.67e-11
M=5.97e24
###POR SIMPLCIDAD PEDIREMOS QUE EL USUARIO INGRESE EL PERIODO EN HORAS Y DESPUES HAREMOS LA CONVERSION A SEGUNDOS.
periodo=float(input("Ingresa el periodo en horas:"))
periodo=periodo*3600
#### Escribimos la ecuacion de la altura y la mostramos en pantalla
altura_s=(((G*M*periodo**2)/(4*pi**2))**(1/3))-R
print("La altura del satelite es:",round(altura_s/1000,2),"kilometros") ## Valor redondeado en kilometros

Ingresa el periodo en horas: 1.5


La altura del satelite es: 279.32 kilometros


(c) Utiliza tu programa para calcular las altitudes de los satélites que orbitan la Tierra una vez al día (la llamada órbita *geoestacionaria*), una vez cada **90 minutos** y una vez cada **45 minutos**. ¿Qué concluyes de este último cálculo?

De acuerdo al programa del inciso (b) las altitudees de los satelites son:

Para un satelite con un perido orbital de 1 día o $24$ horas su altitud de orbita es de $35855.91$ kilometros, para un satelite de periodo orbital de $90$ minutos o $1.5$ horas, la altitud de su orbita es de $279.32$ kilometros, y para un satelite con periodo orbital de $45$minutos o $0.75$ horas, la altitud de su orbita es de $-2181.56km$ lo cual nos dice que para que un objeto pueda tener ese periodo, debe estar dentro de la tierra, por lo que no hay satelite que tenga un perido de 45 minutos

## Resultado 2b y 2c respectivamente: 0.3333/0.3333

¡Muy bien!

# Resultado: 1/1

**3. Relatividad especial:** Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia. 
 
 Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino

a) en el marco de reposo de un observador en la Tierra y 

En este problema al estar en un maro de referencia en reposo, y suponiendo que la na nave espacial se mueve a velocidad constante, el tiempo que mediremos nosotros como observadores en reposo es $t=\frac{d}{v}$

In [3]:
#### Calculamos el tiempo medido desde el observador en reposo.
c=3e8
distancia=float(input("Introduce la distancia en años luz:"))*9.461e15
velocidad=float(input("Introduce la velocidad como una fraccion de la velocidad de la luz, recuerda que esta debe ser menor a 1:"))
if velocidad <= 1:
    tiempo=distancia/(velocidad*c)
    print("El tiempo que mide el observador en reposo es:",round(tiempo/31536000,2),"años")
else:
    print("Nada puede viajar mas rapido que la luz")

Introduce la distancia en años luz: 10
Introduce la velocidad como una fraccion de la velocidad de la luz, recuerda que esta debe ser menor a 1: .99


El tiempo que mide el observador en reposo es: 10.1 años


# Resultado: 0.3/0.5

El resultado esta parcialmente bien, pero no es el resultado que esperamos. De hecho había que usar dilactación temporal tal y como propusiste en el siguiente inciso. El tiempo que percibe un observador en reposo un sistema en movimiento es más lento que el tiempo que percibe alguien dentro de ese sistema (nave)

$$
\Delta t=\gamma\Delta{t}
$$

y por tanto $t>\overline{t}$. Aquí el tiempo del observador en la nave lo podemos determinar fácilmente haciendo $t=\frac{d}{v}$, y esto es posible porque esta es una velocidad bien definida.

b) como lo percibiría un pasajero a bordo de la nave. 

Cuando el observador esta en el sistema en movimiento a una velocidad $\vec{v}$ mide una dilatación temoral dada por las transformaciones de Lorentz, entonces en este caso $t_0=\frac{t}{\sqrt{1-\frac{v²}{c²}}}$ y despejamos el tiempo $t$ que es el tiempo como lo percibiria un pasajero a bordo de la nave desdde el marco de referencia terrestre. $t=t_0\sqrt{1-\frac{v²}{c²}}$

In [3]:
from math import*
c=3e8
distancia_m=float(input("Introduce la distancia en años luz:"))*9.461e15
velocidad_m=float(input("Introduce la velcodiad como una fraccion de la velocidad de la luz, recuerda que esta debe ser menor a 1:"))
if velocidad_m<=1:
    tiempo_m=(distancia_m/(velocidad_m*c))/(sqrt(1-(velocidad_m**2)))
    print("El tiempo que percibec un pasajero de la nave espacial es:",round(tiempo_m/31536000,2),"años")
else:
    print("Nada puede viajar mas rapido que la luz")

Introduce la distancia en años luz:10
Introduce la velcodiad como una fraccion de la velocidad de la luz, recuerda que esta debe ser menor a 1:0.99
El tiempo que percibec un pasajero de la nave espacial es: 71.61 años


Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

De acuerdo a ambos pogramas para el observador pasaran $10.10$ años y para las personas en la nave espacial pasaran $1.43$ años medido desde el marco de referencia en la tierra

## Resultado: 0.3/0.5

Este resultado no puede ser posible porque el observador en movimiento no puede viajar más rápido que la luz, es decir, no puede llegar en menos de 10 años.  Aquí el detalle es que en lugar de dividir la $\gamma$ la pusiste multiplicando por eso no te dio el resultado. Corrigiendo ese detalle observa como el tiempo cambia a 71 años, lo cual sigue siendo incorrecto pero no del todo.

Si me dices cual es la hipótesis que se debe cumplir para que ambos resultados sean correctos, te pongo el punto completo. Es decir, que hipótesis se debe considerar para que el observador en la Tierra perciba 10.10 años y el observador en la nace 71 años

# Resultado: 0.6/1

**4. Órbitas planetarias:** En el espacio, la órbita de un cuerpo alrededor de otro (como un planeta alrededor del Sol), no necesariamente es circular. En general, toma la forma de una elipse, con el cuerpo a veces más cerca y otras más lejos. Si tenemos la distancia $\ell_1$ de máxima aproximación de un planeta al Sol (su *perihelio*), y su velocidad lineal $v_1$ en el perihelio, entonces cualquier otra propiedad de la órbita se puede calcular a partir de estas dos cantidades de la siguiente manera:

   
**a)** La segunda ley de Kepler nos dice que la distancia $\ell_2$ y la velocidad $v_2$ del planeta en su punto más distante, o *afelio*, satisfacen que $\ell_2 v_2 = \ell_1 v_1$.  Al mismo tiempo, la energía total, cinética más la gravitatoria, de un planeta con velocidad $v$ y distancia $r$ del Sol está dada por:

<h3 align="center">$E = \frac{1}{2} m v^2 - G {mM\over r}$,</h3> 
        
donde $m$ es la masa del planeta, $M=1.9891\times10^{30}\,$kg es la masa del Sol y $G=6.6738\times10^{-11}\,\mathrm{m^3\,kg^{-1}\,s^{-2}}$ es la constante gravitatoria de Newton. Dado que la energía debe conservarse, demuestre que $v_2$ es la raíz más pequeña de la ecuación cuadrática:

<h3 align="center">$v_2^2 - {2GM\over v_1\ell_1} v_2 - \biggl[ v_1^2 - {2GM\over\ell_1} \biggr] = 0$.</h3> 

Una vez que tenemos $v_2$ podemos calcular $\ell_2$  usando la relación $\ell_2 = \ell_1 v_1/v_2$.

$\textbf{DEMOSTRACIÓN EN EL PDF DE DEMOSTRACIONES}$

## Resultado: 0.3333/0.3333

**b)** Dados los valores de $v_1$, $\ell_1$, y $\ell_2$; otros parámetros de la órbita se obtienen mediante fórmulas simples que pueden derivarse de las leyes de Kepler y del hecho de que la órbita es una elipse:

<h4 align="center">Semieje mayor: $\; \; \; a = \frac{1}{2}(\ell_1+\ell_2)$,</h4> 

<h4 align="center">Semieje menor: $\; \; \; b = \sqrt{\ell_1\ell_2}$,</h4> 

<h4 align="center">Período orbital: $\; \; \; T = {2\pi ab\over\ell_1 v_1}$,</h4> 

<h4 align="center">Excentricidad orbital: $\; \; \; e = {\ell_2-\ell_1\over\ell_2+\ell_1}$.</h4> 

Escribe un programa que le pida al usuario que ingrese la distancia al Sol y la velocidad en el perihelio; para que calcule e imprima las cantidades:

$\ell_2$, $v_2$, $T$ y $e$.

In [4]:
from math import *
G=6.6738e-11
M=1.9891e30
def v2(v_1,l_1,G,M):
    chicharronera=(2*G*M/(v_1*l_1))+(sqrt((4*G**2*M**2/(v_1**2*l_1*22))+4*(v_1**2-2*G*M/l_1))/(-2*(v_1**2-2*G*M/l_1)))
    return chicharronera
def l2(l_1,v_1,v_2): return(l_1*v_1/v_2)
def semieje_mayor(l_1,l_2): return((1/2)*(l_1+l_2))
def semieje_menor(l_1,l_2): return((sqrt(l_1*l_2)))
def periodo_orbital(l_1,v_1,s_mayor,s_menor):
    a=s_mayor
    b=s_menor
    return ((2*pi*a*b)/(l_1*v_1))
def excentricidad(l_1,l_2): return((l_2-l_1)/(l_1+l_2))

def Orbita(v_1,l_1,G,M):
    v_2=v2(v_1,l_1,G,M)
    l_2=l2(l_1,v_1,v_2)
    sm=semieje_mayor(l_1,l_2)
    s_me=semieje_menor(l_1,l_2)
    po=periodo_orbital(l_1,v_1,sm,s_me)
    print("velocidad v2:",v_2,"m/s")
    print("l2:",l_2,"m")
    print("semi eje mayor:",sm,"m")
    print("semi eje menor:",s_me,"m")
    print("Periodo:",po,"s")
    

**c)** Prueba tu programa haciendo que calcule las propiedades de las órbitas de la Tierra (para las cuales $\ell_1=1.4710\times10^{11}\,$m y $v_1=3.0287\times10^4\,\mathrm{m\,s^{-1}}$)  y del cometa Halley ($\ell_1=8.7830\times10^{10}\,$m y $v_1=5.4529\times10^4\,\mathrm{m\,s^{-1}}$). Entre otras cosas, deberías encontrar que el período orbital de la Tierra es de un año y el del cometa Halley es de unos 76 años.

In [6]:
print("Las propiedades orbitales del cometa Halley:")
Orbita(5.4529e4,8.7830e10,6.6738e-11,1.9891e30)
print("Las propiedades orbitales de la Tierra:")
Orbita(3.0287e4,1.4710e11,6.6738e-11,1.9891e30)

Las propiedades orbitales del cometa Halley:
velocidad v2: 55471.10388025367 m/s
l2: 86338322747.97879 m
semi eje mayor: 87084161373.9894 m
semi eje menor: 87080967420.87204 m
Periodo: 9948826.825911611 s
Las propiedades orbitales de la Tierra:
velocidad v2: 59595.14423808438 m/s
l2: 74758065559.85957 m
semi eje mayor: 110929032779.92978 m
semi eje menor: 104866159669.62527 m
Periodo: 16405577.706088124 s


In [5]:
5.4529e4>55471.10

False

## Resultados ejercicios 4b y 4c respectivamente: 0.3/0.5

Aquí estuviste mal, observa que tu $v_2$ es mayor que la $v_1$ y justo en el inciso anterior habíamos quedado que era al revés. Aquí donde yo encuentro la falla es enla fórmula de la chicharronera, al parecer python se porta mal cuando poner muchas operaciones así de atascadas, es mejor tener la buena práctica de definir variables y ya luego juntarlas en la operación para que se pueda resolver bien. Aunque desde nuestro punto de vista estan bien escritas las operaciones, al parecer python no lo lee muy bien y no se que demonios entrega jaja. Puede que se trate de que hay un error que se va maximizando hasta que crece tanto que diverge del resultado real.

Otro detalle es que para el periodo orbital había que dividir el periodo encontrado entre el número de segundos que tiene un año y corroborar que son de 1 año y 76 años.

# Resultado: 0.96/1

**5. La fórmula semiempírica de la masa (FSM)**
 
En física nuclear, la **fórmula de Weizsäcker** (conocida también como fórmula semiempírica) sirve para evaluar la masa y otras propiedades de un núcleo atómico; y está basada parcialmente en mediciones empíricas. En particular la fórmula se usa para calcular la $\textbf{\textit{energía de enlace nuclear aproximada}}$ $B$, de un núcleo atómico con número atómico $Z$ y número de masa $A$:

<h3 align="center">$B = a_1 A - a_2 A^{2/3} - a_3 {Z^2\over A^{1/3}} - a_4 {(A - 2Z)^2\over A} + {a_5\over A^{1/2}}$,</h3>

donde, en unidades de millones de electrón-volts, las constantes son $a_1=15.8$, $a_2=18.3$, $a_3=0.714$, $a_4=23.2$ y          
<h6 align="center">
$a_5 = 
  \left \{
    \begin{aligned}
      0 & \; \text{si} \ A \; \text{es impar,} \\
      12.0 & \; \text{si} \ A \; \text{y} \; Z \; \text{son pares (ambos),} \\
      -12.0 & \; \text{si} \ A \; \text{es par y} \; Z \; \text{impar} 
    \end{aligned}
  \right.
$
</h6>

**a)** Escribe un programa que tome como entrada los valores de $A$ y $Z$, e imprima la energía de enlace $B$ para el átomo correspondiente. Usa tu programa para encontrar la energía de enlace de un átomo con $A=58$ y $Z=28$ (Hint: La respuesta correcta es alrededor de los $490\,$MeV).

In [6]:
def energia_atomo(A,Z):
    a1=15.8
    a2=18.3
    a3=0.714
    a4=23.2
    a5=0
    if A%2 == 0 and Z%2 == 0:
        a5=12
    elif A%2 == 0 and Z%2!=0:
        a5=-12
    elif A%2 != 0:
        a5=0
    B=a1*A-a2*A**(2/3)-a3*(Z**2/A**(1/3))-a4*((A-2*Z)**2/A)+a5/A**(1/2)
    return B

In [7]:
print("La energia de enlace de A=58 y Z=28 es:", energia_atomo(58,28),"Mev")

La energia de enlace de A=58 y Z=28 es: 497.5620206224374 Mev


## Resultado: 0.25/0.25

**b)** Modifica el programa del inciso anterior, para escribir una segunda versión que imprima no la energía de enlace total $B$, sino la energía de unión por nucleón, que es $B/A$.

In [8]:
def energia_nucleon(A,Z):
    B=energia_atomo(A,Z)
    return B/A

## Resultado: 0.25/0.25

Muy bien, reciclando funciones esta genial. Ahorras espacio y optimizas código.

In [9]:
print("La energia de union por nucleon para A=58,Z=28 es:",energia_nucleon(58,28),"Mev")

La energia de union por nucleon para A=58,Z=28 es: 8.578655527973059 Mev


**c)** Escribe una tercera versión del programa para que tome como entrada solo un valor del número atómico $Z$ y luego pase por todos los valores de $A$ desde $A=Z$ hasta $A=3Z$, para encontrar el que tiene la mayor energía de enlace por nucleón . Este es el núcleo más estable con el número atómico dado. Haz que tu programa imprima el valor de $A$ para este núcleo más estable y el valor de la energía de enlace por nucleón.

In [10]:
def mayor_energia_nucleon(Z):
    import numpy as np
    a=list(np.arange(Z,(3*Z)+1,1))
    energia=[]
    A=[]
    for i in a:
        energia.append(energia_nucleon(i,Z))
        A.append(i)
    m_energia=max(energia)
    atomo=A[energia.index(m_energia)]
    return [atomo,m_energia]

In [11]:
mayor_energia_nucleon(28)

[62, 8.70245768367189]

# Resultado: 0.25/0.25

Aquí tuviste un error (que ya corregí) muy minúsculo, pero que venías arrastrando desde el primer inciso. Lo que pasó es que en las condiciones donde $a_5=\pm 12$ olvidaste poner la condición `A%2 == 0`; ese pequeño detalle te estaba arrohando un pequeño sesgo, te salía la respuesta $A=63$ y la correcta es $A=62$ igual te aparecía otra energía que no era la correspondiente. Pero bueno, arreglando ese detalle no hay mayor problema, tu código esta perfecto y eres de los pocos que usó listas para atacar el problema, ¡bien ahi!

In [12]:
print("El numero A para el cual el nucleo es mas estable es:",mayor_energia_nucleon(28)[0],"Y el valor de la mayor energia de enlace por nucleon es:",mayor_energia_nucleon(28)[1],"Mev")

El numero A para el cual el nucleo es mas estable es: 62 Y el valor de la mayor energia de enlace por nucleon es: 8.70245768367189 Mev


**d)** Finalmente, escribe una cuarta versión del programa que, en lugar de tomar $Z$ como entrada, se ejecute a través de todos los valores de $Z$ de 1 a 100 e imprima el valor más estable de $A$ para cada uno. ¿A qué valor de $Z$ se produce la energía de enlace máxima por nucleón? (La respuesta correcta, en la vida real, es $Z=28$, que corresponde al Níquel).

In [15]:
import numpy as np
Z=np.arange(1,101,1)
estabilidad=[]
max_energia_nucleon=[]
for i in range(1,100,1):
    estabilidad.append(mayor_energia_nucleon(i)[0])
    max_energia_nucleon.append(mayor_energia_nucleon(i)[1])
##AQUI IMRPIMIMOS LOS VALORES DE A PARA LOGRAR LA ESTABLIDAD EN NUMEROS ATOMICOS DEL 1 AL 100
print("Valores de A para los cuales el atomo es estable:")
print(estabilidad)
### AQUI PRODUCIMOS EL VALOR DE A PARA UNA MAXIMA DE ENERGIA DE ENLACE POR NUCLEON 
print("Valor de A para el cual la energia de enlace por nucleon es maxima:")
print(estabilidad[max_energia_nucleon.index(max(max_energia_nucleon))])
### AQUI IMPRIMIOS A QUE VALOR DE z en Z corresponde ese A
print("Valor de Z para la cual se produce la energía de enlace máxima por nucleon para A:")
print(Z[max_energia_nucleon.index(max(max_energia_nucleon))])
print(max(max_energia_nucleon))

Valores de A para los cuales el atomo es estable:
[3, 4, 7, 8, 11, 14, 15, 18, 19, 22, 25, 26, 29, 30, 33, 36, 37, 40, 43, 44, 47, 48, 51, 54, 55, 58, 61, 62, 65, 68, 69, 72, 75, 76, 79, 82, 85, 86, 89, 92, 93, 96, 99, 102, 103, 106, 109, 110, 113, 116, 119, 120, 123, 126, 129, 130, 133, 136, 139, 140, 143, 146, 149, 150, 153, 156, 159, 162, 163, 166, 169, 172, 175, 176, 179, 182, 185, 188, 191, 192, 195, 198, 201, 204, 205, 208, 211, 214, 217, 220, 223, 224, 227, 230, 233, 236, 239, 242, 243]
Valor de A para el cual la energia de enlace por nucleon es maxima:
62
Valor de Z para la cual se produce la energía de enlace máxima por nucleon para A:
28
8.70245768367189


## Resultado: 0.25/0.25

Excelente, volviste a usar la lógica del inciso anterior pero extendida y te salió perfecto. Agregué una línea a tu código, el último print para saber cual es la energía correspondiente y coincide con el inciso anterior, muy bien.

# Resultado: 1/1

# Resultado Final:

* Ejercicio 1: 1
* Ejercicio 2: 1
* Ejercicio 3: 0.6
* Ejercicio 4: 0.96
* Ejercicio 5: 1

**Total: 9.1**

